### PR TITLE
feat: impl `__field__` special matcher to project value columns

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Implement `__field__` special matcher. It can do projection on value columns under multi-value model. E.g.:

```sql
MySQL [(none)]> TQL eval (1667446797, 1667446807, '10s') system_metrics{ __field__="cpu_util"};
+----------+-------+-------+---------------------+
| cpu_util | host  | idc   | ts                  |
+----------+-------+-------+---------------------+
|     11.8 | host1 | idc_a | 2022-11-03 03:40:07 |
|       50 | host1 | idc_b | 2022-11-03 03:40:07 |
|     80.1 | host2 | idc_a | 2022-11-03 03:40:07 |
+----------+-------+-------+---------------------+
3 rows in set (0.005 sec)

MySQL [(none)]> TQL eval (1667446797, 1667446807, '10s') system_metrics{ __field__="cpu_util", __value__="disk_util"};
+-----------+----------+-------+-------+---------------------+
| disk_util | cpu_util | host  | idc   | ts                  |
+-----------+----------+-------+-------+---------------------+
|      10.3 |     11.8 | host1 | idc_a | 2022-11-03 03:40:07 |
|      40.6 |       50 | host1 | idc_b | 2022-11-03 03:40:07 |
|        90 |     80.1 | host2 | idc_a | 2022-11-03 03:40:07 |
+-----------+----------+-------+-------+---------------------+
3 rows in set (0.004 sec)

MySQL [(none)]> TQL eval (1667446797, 1667446807, '10s') system_metrics{ __field__!="cpu_util"};
+-----------+-------------+-------+-------+---------------------+
| disk_util | memory_util | host  | idc   | ts                  |
+-----------+-------------+-------+-------+---------------------+
|      10.3 |        10.3 | host1 | idc_a | 2022-11-03 03:40:07 |
|      40.6 |        66.7 | host1 | idc_b | 2022-11-03 03:40:07 |
|        90 |        70.3 | host2 | idc_a | 2022-11-03 03:40:07 |
+-----------+-------------+-------+-------+---------------------+
3 rows in set (0.005 sec)
```

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/promql/src/planner.rs [hunk 1]
```diff
@@ -431,10 +441,78 @@ impl PromPlanner {
...
+                    MatchOp::NotEqual => {
+                        if col_set.contains(&matcher.value) {
+                            reverse_set.insert(matcher.value.clone());
+                        } else {
+                            return Err(ValueNotFoundSnafu {
+                                table: self.ctx.table_name.clone().unwrap(),
+                            }
+                            .build());
+                        }
+                    }
```
Comment (yuanbohan) [inline src/promql/src/planner.rs:476]: for `Equal` and `NotEqual`, which is the behavior of Prometheus, empty result or error?

File: src/promql/src/planner.rs [hunk 2]
```diff
@@ -431,10 +441,78 @@ impl PromPlanner {
...
+                                reverse_set.insert(col.clone());
+                            }
+                        }
+                    }
+                }
+            }
+            // merge two set
+            if result_set.is_empty() {
+                result_set = col_set.into_iter().cloned().collect();
+            }
```
Comment (yuanbohan) [inline src/promql/src/planner.rs:496]: empty result means all columns. Is this the behavior of Prometheus?

File: src/promql/src/planner.rs [hunk 3]
```diff
@@ -431,10 +441,78 @@ impl PromPlanner {
...
             .create_table_scan_plan(&table_name, filters.clone())
             .await?;
 
+        // make a projection plan if there is any `__value__` matcher
+        if let Some(value_matchers) = &self.ctx.value_column_matcher {
+            let col_set = self.ctx.value_columns.iter().collect::<HashSet<_>>();
+            // opt-in set
+            let mut result_set = HashSet::new();
+            // opt-out set
+            let mut reverse_set = HashSet::new();
```
Comment (yuanbohan) [inline src/promql/src/planner.rs:454]: Is it possible to use just one mutable set, like `target_set`?

File: src/promql/src/planner.rs (hunk 1 above)
Comment (waynexia) [reply src/promql/src/planner.rs:476]: PromQL doesn't throw error in most cases like wrong table name, wrong column name etc. But we need to report error for better UE. I plan to introduce a config like "prom_compatible_mode" to switch this behavior. But in this case `__field__` doesn't exist in promql. It's non-"prom_compatible_mode" only feature so I just throw an error.

Comment (yuanbohan) [reply src/promql/src/planner.rs:476]: What about `Re` and `NotRe`, Is there a chance to throw error? It's like `Equal` and `Re` behaves differently.

Comment (waynexia) [reply src/promql/src/planner.rs:476]: They are not accurate matches, so I ignore the unmatching case for regex.

File: src/promql/src/planner.rs [hunk 4]
```diff
@@ -1690,4 +1772,125 @@ mod test {
 
         indie_query_plan_compare(query, expected).await;
     }
+
+    #[tokio::test]
+    async fn value_matcher() {
```
Comment (yuanbohan) [inline src/promql/src/planner.rs:1777]: what about some error scenarios?

**Comment by killme2008** [comment]
@waynexia The PR content must be edited too. The `__value__` is renamed to `__field__`.

File: src/promql/src/planner.rs [hunk 5]
```diff
@@ -76,6 +80,7 @@ struct PromPlannerContext {
     time_index_column: Option<String>,
     value_columns: Vec<String>,
     tag_columns: Vec<String>,
+    value_column_matcher: Option<Vec<Matcher>>,
```
Comment (killme2008) [inline src/promql/src/planner.rs:83]:
```suggestion
    field_column_matcher: Option<Vec<Matcher>>,
```

**Review by yuanbohan — APPROVED** [review]

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**